metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -1652,9 +1652,7 @@ private void registerWriteNoOpRecord(long maxIdleIntervalNs) {
         periodicControl.registerTask(new PeriodicTask("writeNoOpRecord",
             () -> {
                 ArrayList<ApiMessageAndVersion> records = new ArrayList<>(1);
-                if (featureControl.metadataVersion().isNoOpRecordSupported()) {
-                    records.add(new ApiMessageAndVersion(new NoOpRecord(), (short) 0));
-                }
+                records.add(new ApiMessageAndVersion(new NoOpRecord(), (short) 0));
```

Comment:
Maybe we can streamline the code?
```
() -> ControllerResult.of(List.of(new ApiMessageAndVersion(new NoOpRecord(), (short) 0)), false),
```

Reply:
The new version would make `records` immutable while the current code exposes the mutable `records` via `ControllerResult.records()`. Do you know if this change is safe? My cursory look indicates it may be ok, but I would have to look at this more carefully. Given that, I think we should also do this via a different PR (to avoid including a potential behavior change in the middle of this clean-up PR).

Reply:
> Do you know if this change is safe? My cursory look indicates it may be ok, but I would have to look at this more carefully. 

Since other similar cases pass immutable lists to `ControllerResult`, it seems safe to do so here as well.

> I think we should also do this via a different PR (to avoid including a potential behavior change in the middle of this clean-up PR).

make sense to me

Reply:
@ijuma , we do not mutate the records list that gets returned from these functions later on. @chia7712 is correct that we return immutable objects in many other cases. Unfortunately, due to Java's lack of `const` I guess that isn't obvious.

Reply:
OK, I'll include this change in my next push.